### ISSUE TITLE
engine: mlx-vlm exact-verifier falls back to per-token linears on mxfp targets — spec decoding capped at ~1.0x (root cause + fix plan)

### ISSUE THREAD

**Issue by krcm0209** [krcm0209/sous#58]
Follow-up to #55. The "speculative decoding gives ~1.0x despite 94.5% acceptance" result is now fully root-caused: it is an mlx-vlm 0.6.16 bug specific to mxfp-quantized targets, with a verified 3-line local fix and a scoped upstream plan. Filing so the decision (patch vs affine target vs upstream PRs) can be made later. Nothing has been filed upstream yet.

Environment for all numbers: M5 Pro / 64 GB, macOS 26.6.2, mlx 0.32.2, mlx-vlm 0.6.16, `mlx-community/Qwen3.8-27B-mxfp8` + `z-lab/Qwen3.8-27B-DFlash2` (4-bit in-memory), greedy, 160 tokens, measured 2026-08-29.

## Root cause

With `draft_model` set, the round loop calls the target with `speculative_verify=True` (`mlx_vlm/speculative/dflash.py:313`), which reroutes the whole verify forward through `Qwen3_5ExactSpeculativeVerifier` (`mlx_vlm/models/qwen3_5/language.py:2051`) — a path whose job is to make block verification bit-identical to single-token decode so speculative output exactly matches the no-drafter baseline.

Its fast "exact" quantized kernels gate on **affine** mode, bits ∈ {4,5,8}, affine `biases` present, bf16/fp16 scales (`_can_target_verify_quantized_head`, `mlx_vlm/models/qwen3_5/speculative_verifier.py:908`). Every layer of an mxfp8/mxfp4 checkpoint is `(mode='mxfp8', bits=8, biases=None, uint8 scales)` and fails the gate, so all **497** QuantizedLinears silently fall into `_target_verify_timewise` (`speculative_verifier.py:1073`):

```python
def _target_verify_timewise(fn, x):
    return mx.concatenate([fn(x[:, i:i+1]) for i in range(x.shape[1])], axis=1)
```

One single-token matmul per verify position per linear. A 27B decode forward is memory-bound on weight reads, so a width-3 verify re-reads all 28.66 GB ~3x.

## Evidence

Width-3 forward, same warm cache, forced eval + sync fences, best of 4:

| path | T=3 forward |
|---|---:|
| normal forward (no flag) | 105.6 ms |
| `speculative_verify=True`, stock | **289.0 ms** |
| same, linear helpers patched to batched | 107.6 ms |
| normal forward, T=1 (cross-check vs 9.9 tok/s baseline) | 100.8 ms |

Counters during the stock call: `_target_verify_timewise` fired 497x -> 1,491 single-token matmuls per verify; the exact kernels fired 0 times. Accounting closes: 289 verify + 10.6 drafter + ~4 snapshot/walk ≈ the observed 304 ms round (16.73 s / 55 rounds). The GDN state checkpointing and per-position attention are *not* the cost (row 3).

End-to-end (all runs token-identical to baseline; acceptance unchanged by the patch):

| run | tok/s | ratio |
|---|---:|---:|
| baseline | 9.90 | 1.00x |
| stock spec, block=3 | 9.68 | 0.98x |
| patched, block=3 | 24.04 | **2.43x** |
| patched, auto block | 30.64 | **3.10x** |

The auto block policy flips from worst (0.83x stock — every extra drafted token bought a full weight sweep) to best. The earlier mxfp4 flat results (0.99–1.05x) are the same bug: predicted round cost ≈ T x 58.8 ms matches what was measured. Upstream's 1.75x claim (Blaizzy/mlx-vlm PR #2014) is genuine and consistent — it used `Qwen3.8-27B-4bit`, which is affine and takes the fast kernels (verified directly against `_target_verify_quantized_linear` with synthetic layers: affine-4/8 with bf16 scales -> kernel; mxfp8/mxfp4 -> timewise).

## Local fix (verified)

```python
import mlx_vlm.models.qwen3_5.speculative_verifier as sv
sv._target_verify_linear = lambda linear, x: linear(x)
sv._target_verify_linears = lambda linears, x: tuple(l(x) for l in linears)
sv._target_verify_embedding_as_linear = lambda emb, x: emb.as_linear(x)
```

Caveat: this trades the bit-exactness guarantee for MLX's batched kernels. Measured bitwise: MLX's batched quantized matmul is **not** row-stable vs singleton for any mode (mxfp8/mxfp4/affine, T=2–64, square and head-shaped), so the patch is "practically exact" (token-exact on the gate prompt), not exact by construction. Worst case of an fp-noise argmax flip: one differing-but-legitimate committed token or one rejected draft — never corruption, since committed tokens always come from target logits.

## Decision for sous (later)

1. **Affine target, zero code**: use an affine checkpoint (e.g. `Qwen3.8-27B-4bit`, or affine 8-bit ≈ mxfp8 size class) as the spec-decoding target — the exact kernels engage today. Needs a quality spot-check vs mxfp on the tool loop.
2. **Carry the 3-line patch**, pinned to mlx-vlm 0.6.16 with a comment, accepting practical-not-guaranteed exactness (immaterial for mechanical codegen).
3. **Upstream it** (below) and wait.

Either way the #55 economics flip: ~30 tok/s local decode at 94.5% acceptance, and mxfp4 + spec decoding likely stack further (unverified).

## Upstream plan (Blaizzy/mlx-vlm), if pursued

**PR 1 — the silent cliff (small, immediate value):** one-time warning (or opt-in error) when `_target_verify_timewise`/`_target_verify_singletons` engages, plus a documented `exact_verify=False` flag plumbed from `generate()`/server config that routes the three linear helpers to plain batched ops (distribution-preserving, not bit-exact). Also covers LFM2, whose verifier has the same failure class for *any* quantized target (its fast kernel requires dense bf16/fp16 weights). ~Half a day + plumbing/docs.

**PR 2 — mxfp exact kernels (the proper fix, ~3–5 focused days):**
- Recon mlx-core's mxfp qmv in `mlx/backend/metal/kernels/quantized.h` — "exact" means bit-matching MLX's singleton kernel, so decode + accumulation order must be mirrored. The affine exact kernels mirror the affine qmv skeleton (32-lane SIMD, 2 simdgroups, 4 results/simdgroup, 2 packs/thread); if mxfp differs structurally, the header template forks (+1 day).
- Header `_target_verify_qlinear_header`: mode branches in `qdot_exact` — mxfp4 = nibble unpack + 16-entry e2m1 LUT (confirmed: MLX mxfp4 elements are e2m1); mxfp8 = e4m3 decode (confirmed); both drop the affine `sum*bias` trick and decode uint8 e8m0 scales (group 32, no biases).
- Variants: base QMV + QARGMAX required for block ≤ 5 (masked argmax is string-derived, comes free); token-tiled variants required for auto-block (T up to ~16, register pressure caps the base kernel); streamed and fused-qkv variants are optional polish.
- Dispatch: extend the `_can_target_verify_quantized*` gates for mxfp constraints, add `mode` to kernel cache keys and variant selection.
- Tests: `mlx_vlm/tests/test_speculative.py` already asserts bitwise kernel-vs-`_target_verify_timewise` equality — parametrize over mode x dtype x T(1–16) x shapes (square, head-shaped, K%512 boundary). Metal-gated, local-run like our model tests.
- Finicky bits (each shows up as a red bitwise test): fp32 accumulation order, the deliberate `float(T(simd_sum(...)))` store-rounding in argmax kernels, e4m3 NaN slot, denormals.
- Risks: maintainer may redirect the work into mlx-core ("row-deterministic batched qmm belongs upstream"); the exactness contract permanently chases mlx-core kernel changes (already true for affine).

**Separate tiny PR:** `RadixArk/Qwen3.8-27B-DSpark` cannot load on stock 0.6.16 — `mlx_vlm/speculative/drafters/dspark/config.py:114` asserts `hidden_size == heads * head_dim`, which decoupled-head_dim checkpoints legitimately violate (weights confirm 32x128=4096 projecting back to 5120; the target itself breaks the same identity). Deleting the 4-line check suffices: loads, validates, token-exact output, 4 failed / 2578 passed on `mlx_vlm/tests/` with the 4 failures pre-existing on unpatched v0.6.16.

## Notes

- Both drafters (DFlash2, DSpark) hit the same tax — it is drafter-independent.
- Measurement discipline that mattered: mlx is lazy and `mx.synchronize()` does not evaluate an un-scheduled graph; every number above forces eval on returned arrays and cross-checks against end-to-end wall clock.

**Comment by krcm0209** [comment]
**Option 1 verified (performance half): affine-4bit + stock mlx-vlm passes the gate.**

Pulled `mlx-community/Qwen3.8-27B-4bit` and reran the spec-decoding gate, stock 0.6.16, no patch — same prompt/harness as the mxfp8 numbers above (M5 Pro, greedy, 160 tok, DFlash2 drafter 4-bit in-memory):

| run | tok/s | ratio | exact |
|---|---:|---:|---|
| baseline | 17.26 | 1.00x | — |
| spec block=2 | 27.11 | 1.57x | yes |
| spec block=3 | 38.49 | 2.23x | yes |
| spec block=5 | 38.73 | 2.24x | yes |
| spec auto | **40.79** | **2.36x** | yes |

- Dispatch counters: `timewise = 0` in every run; the affine exact kernels fired (~7k qmv hits per block=3 run). The fallback never engages on this checkpoint.
- Acceptance 98.1% at block=3 (slightly above mxfp8's 94.5%); auto settles at ~3.9 drafts/round.
- Baseline is ~6% under mxfp4's 18.3 tok/s — matches the 4.5 vs 4.25 bpw arithmetic.
- Beats the patched-mxfp8 route (30.6 tok/s) in absolute throughput while keeping the bit-exactness guarantee, and beats upstream's published 1.75x.

**Net: the monkeypatch (option 2) is obsolete for sous — option 1 wins on speed, cleanliness, and guarantees.** Remaining before switching the default model: the quality half — one real delegated tool-loop task A/B, 4bit vs mxfp8.

**Comment by krcm0209** [comment]
**Option 1 verified (quality half): tool-loop A/B shows no degradation on 4bit.**

Ran 3 graded delegation tasks through the real worker loop (`run_worker_loop`, tmp dirs, shipped sampling temp 0.7, default allowlist) per model: module-from-spec graded against hidden tests, a docstring sweep graded by AST + behavior checks, and test scaffolding graded against the pristine module.

| task | 4bit | mxfp8 |
|---|---|---|
| t1 module-from-spec | completed, 4 turns, 80s — hidden tests 8/8, own tests 10/10 | budget-exhausted, 30 turns, 476s — hidden tests 8/8, own tests 10/11 |
| t2 docstring sweep | completed, 7 turns, 300s — perfect artifact | budget-exhausted at 10-min wall, 11 turns — artifact also perfect |
| t3 test scaffolding | completed, 3 turns, 30s — 5/5 vs pristine | completed, 5 turns, 90s — 6/6 vs pristine |
| **artifact score** | **6/6** | **5/6** |
| total wall / turns | 411s / 14 | 1167s / 46 |

Notes for honest reading:
- N=1 per cell at temp 0.7 — this is a **parity check for 4bit, not evidence mxfp8 is dumber**. mxfp8's one miss was a wrong self-authored test expectation (asserted a 10-char string should middle-truncate at max_len=30); its module was correct, and it burned the turn budget fighting its own red test. Plausibly sampling variance.
- The operational finding is real regardless: at 0.57x decode speed, mxfp8 can't fit the same trajectories inside fixed wall budgets — t2 produced a perfect artifact but still ended budget-exhausted.

**Verdict: quality gate passes. Recommend switching the default model to `mlx-community/Qwen3.8-27B-4bit`** — no observed quality loss on representative delegation work, ~1.7x baseline decode, and speculative decoding works stock at up to 40.8 tok/s (2.36x) with bit-exactness intact. Both halves of option 1 are now verified; options 2 (monkeypatch) can be dropped, and option 3 (upstream PRs) remains worthwhile for the ecosystem but is no longer load-bearing for sous.

**Comment by krcm0209** [comment]
Done in effect: #59 switched the default model to mlx-community/Qwen3.8-27B-4bit (both halves of option 1 verified above) and #60 shipped speculative decoding on the VLM backend, where the affine checkpoint passes the exact-verifier gate stock. Option 2 (the monkeypatch) is obsolete; option 3 (upstream PRs for the mxfp gate) was never filed and is not needed for sous.
